dash_auth/oidc_auth.py

```diff
@@ -0,0 +1,298 @@
+        self.idp_selection_route = idp_selection_route
+        self.logout_page = logout_page
+
+        if secret_key is not None:
+            app.server.secret_key = secret_key
+
+        if app.server.secret_key is None:
+            raise RuntimeError(
+                """
+                app.server.secret_key is missing.
+                Generate a secret key in your Python session
+                with the following commands:
+                >>> import os
+                >>> import base64
+                >>> base64.b64encode(os.urandom(30)).decode('utf-8')
+                and assign it to the property app.server.secret_key
+                (where app is your dash app instance), or pass is as
+                the secret_key argument to OIDCAuth.__init__.
+                Note that you should not do this dynamically:
+                you should create a key and then assign the value of
+                that key in your code/via a secret.
+                """
+            )
+
+        super().__init__(app)
+
+        self.oauth = OAuth(app.server)
+
+        # Check that the login and callback rules have an <idp> placeholder
+        if not re.findall(r"/<idp>(?=/|$)", login_route):
+            raise Exception(
+                "The login route must contain a <idp> placeholder."
+            )
+        if not re.findall(r"/<idp>(?=/|$)", callback_route):
+            raise Exception(
+                "The callback route must contain a <idp> placeholder."
+            )
+
+        app.server.add_url_rule(
+            login_route,
+            endpoint="oidc_login",
+            view_func=self.login_request,
+            methods=["GET"],
+        )
+        app.server.add_url_rule(
+            logout_route,
+            endpoint="oidc_logout",
+            view_func=self.logout,
+            methods=["GET"],
+        )
+        app.server.add_url_rule(
+            callback_route,
+            endpoint="oidc_callback",
+            view_func=self.callback,
+            methods=["GET"],
+        )
+
+    def register_provider(self, idp_name: str, **kwargs):
+        """Register an OpenID Connect provider.
+
+        :param idp_name: The name of the provider
+        :param kwargs: Keyword arguments passed to OAuth.register.
+            See https://docs.authlib.org/en/latest/client/flask.html for
+            additional details.
+            Typical keyword arguments for OIDC include:
+            * client_id
+            * client_secret
+            * server_metadata_url
+            * token_endpoint_auth_method
+            * client_kwargs (defaults to {"scope": "openid email"})
+        """
+        client_kwargs = kwargs.pop("client_kwargs", {})
+        client_kwargs.setdefault("scope", "openid email")
+        self.oauth.register(
+            idp_name, client_kwargs=client_kwargs, **kwargs
+        )
+
+    def get_oauth_client(self, idp: str):
+        """Get the OAuth client."""
+        if not idp:
```

Comment:
can idp be undefined at all in these two get function?

Reply:
Not if it's called from `login_request` but a user might be calling this somewhere else

Reply:
Fair. I think we should even test that idp is a valid one here.
```suggestion
        if not idp in self.oauth._registry:
```